packages/gatsby-worker/src/index.ts

```diff
@@ -0,0 +1,322 @@
+  ...args: Array<any>
+) => Promise<any>
+  ? MaybeFunction
+  : MaybeFunction extends (...args: Array<any>) => any
+  ? WrapReturnOfAFunctionInAPromise<MaybeFunction>
+  : never
+
+type WrapReturnInArray<MaybeFunction> = MaybeFunction extends (
+  ...args: Array<any>
+) => any
+  ? (...a: Parameters<MaybeFunction>) => Array<ReturnType<MaybeFunction>>
+  : never
+
+export type CreateWorkerPoolType<ExposedFunctions> = WorkerPool &
+  {
+    [FunctionName in keyof ExposedFunctions]: EnsureFunctionReturnsAPromise<
+      ExposedFunctions[FunctionName]
+    >
+  } & {
+    all: {
+      [FunctionName in keyof ExposedFunctions]: WrapReturnInArray<
+        EnsureFunctionReturnsAPromise<ExposedFunctions[FunctionName]>
+      >
+    }
+  }
+
+const childWrapperPath = require.resolve(`./child`)
+
+class TaskInfo<T> {
+  functionName: T
+  args: Array<any>
+  assignedToWorker?: IWorkerInfo<T>
+  promise: Promise<any>
+  resolve!: (r: any) => void
+  reject!: (e: Error) => void
+
+  constructor(opts: {
+    functionName: T
+    args: Array<any>
+    assignedToWorker?: IWorkerInfo<T>
+  }) {
+    this.functionName = opts.functionName
+    this.args = opts.args
+    this.assignedToWorker = opts.assignedToWorker
+    this.promise = new Promise((resolve, reject) => {
+      this.resolve = resolve
+      this.reject = reject
+    })
+  }
+}
+
+interface IWorkerInfo<T> {
+  workerId: number
+  worker: ChildProcess
+  exitedPromise: Promise<{
+    code: number | null
+    signal: NodeJS.Signals | null
+  }>
+  currentTask?: TaskInfo<T>
+}
+
+export class WorkerPool<WorkerModuleExports = Record<string, unknown>> {
+  /**
+   * Schedule task execution on all workers. Useful for setting up workers
+   */
+  all: {
+    [FunctionName in keyof WorkerModuleExports]: WrapReturnInArray<
+      EnsureFunctionReturnsAPromise<WorkerModuleExports[FunctionName]>
+    >
+  }
+
+  /**
+   * Schedule task execution on single worker. Useful to distribute tasks between multiple workers.
+   */
+  single: {
+    [FunctionName in keyof WorkerModuleExports]: EnsureFunctionReturnsAPromise<
+      WorkerModuleExports[FunctionName]
+    >
+  }
+
+  private workers: Array<IWorkerInfo<keyof WorkerModuleExports>> = []
+  private taskQueue = new TaskQueue<TaskInfo<keyof WorkerModuleExports>>()
+  private idleWorkers: Set<IWorkerInfo<keyof WorkerModuleExports>> = new Set()
+
+  constructor(workerPath: string, options?: IWorkerOptions) {
+    const single: Partial<WorkerPool<WorkerModuleExports>["single"]> = {}
+    const all: Partial<WorkerPool<WorkerModuleExports>["all"]> = {}
+
+    {
+      // we don't need to retain these
+      const module: WorkerModuleExports = require(workerPath)
+      const exportNames = Object.keys(module) as Array<
+        keyof WorkerModuleExports
+      >
+
+      for (const exportName of exportNames) {
+        if (typeof module[exportName] !== `function`) {
+          // we only expose functions
```

Comment:
Short answer is that it would require more work to support and we only really want to be able to execute functions in worker (heavy work) and not really get access to other exports - if user want access to other exports they should just require that module directly. WorkerPool is only there to handle triggering function execution in worker and reporting back results. In any case added https://github.com/gatsbyjs/gatsby/pull/32120/commits/102007a2000684cc7c9dae8e6d9d10ee614d192e